src/libs/protoboard/plugin.h

```diff
@@ -0,0 +1,55 @@
+
+/***************************************************************************
+ * Protoboard plugin template
+ * - Main plugin template: Instantiate this with the appropriate interface
+ *   handler mappings to create a domain-specific plugin.
+ *
+ * Copyright 2019 Victor Mataré
+ ****************************************************************************/
+
+/*  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Library General Public License for more details.
+ *
+ *  Read the full text in the LICENSE.GPL file in the doc directory.
+ */
+
+#ifndef PROTOBOARD_PLUGIN_H
+#define PROTOBOARD_PLUGIN_H
+
+#include "blackboard_manager.h"
+#include "protobuf_thread.h"
+
+#include <core/plugin.h>
+
+/**
+ * The main class template that generates a domain-specific plugin
+ * @tparam IfaceManagerTs A list of @a bb_iface_manager instantiations that specify
+ * what message types should be handled on a given interface type.
+ */
+template <class... IfaceManagerTs>
+class ProtoboardPlugin : public fawkes::Plugin
+{
+public:
+	/**
+	 * Initializes all threads required for the plugin.
+	 * @param cfg The fawkes config
+	 */
+	ProtoboardPlugin(fawkes::Configuration *cfg) : Plugin(cfg)
+	{
+		protoboard::ProtobufThead *    protobuf_thread = new protoboard::ProtobufThead();
+		protoboard::BlackboardManager *bb_mgr = new protoboard::BlackboardManager(protobuf_thread);
+		bb_mgr->set_protobuf_sender(new protoboard::ProtobufSender<IfaceManagerTs...>(bb_mgr));
+		protobuf_thread->set_bb_manager(bb_mgr);
```

Comment:
Well, I'm aware and I do generally try to avoid these situations. But in this case there's no easy way around it, except maybe apart from a major redesign.

Reply:
The simple explanation is that the control flows both ways: From the `ProtobufThread` to the `BlackboardManager` on incoming messages, and the other way on outgoing messages.